### PR TITLE
[CLR Profiler] Fix InvalidProgramException: Ensure we always run argument-loading instructions before calling instrumentation wrapper methods, even after branches

### LINKED ISSUE #541: InvalidProgramException after successful MySQL connection
**Describe the bug**
After a successful MySQL connection is initiated, the program crashes with an InvalidProgramException inside `MySql.Data.MySqlClient.Driver.Configure(MySqlConnection connection)` if the DataDog Tracer MSI is installed on the host machine.

**To Reproduce**
I have provided a repro here: https://github.com/ksjoberg/dd-trace-repro1
I set this up in IIS, but I believe also regular/console applications are affected.

**Expected behavior**
The tracer/agent should instrument the assemblies transparently.

**Runtime environment:**
- Instrumentation mode: automatic with msi installer
- Tracer version: 1.8.0
- OS: Windows 10 build 1903
- CLR: .NET Framework 4.8

**Additional context**
This is obviously major issue for us.

### PR DESCRIPTION
**Pull request by zacharycmontoya**
Fixes #541

The issue:
IL methods may have jump or branch instructions that go directly to the IL instruction that calls the target method we intend to instrument. Previously, we would replace the metadata token in that call instruction to the MethodRef token that points to our instrumentation wrapper method and we would insert IL instructions before the method call to load additional arguments we need for proper instrumentation. But in this particular branching scenario, we encounter an `InvalidProgramException` because the branching operation jumped past the argument-loading instructions and straight to the method call.

The fix:
Make the original method call instruction a no-op, add the argument-loading instructions after this instruction, and finally add an instruction to call our wrapper method. This ensures the above scenario cannot happen by guaranteeing that argument-loading instructions will be executed before calling the instrumentation wrapper method.

Changes proposed in this pull request:
- Add the above fix in the profiler
- Add the bad code path to an existing and running reproduction case

@DataDog/apm-dotnet

### CODE REVIEW

**Comment by zacharycmontoya** [comment]
Evidence of the failing test case EntityFramework6x.MdTokenLookupFailure when this isn't fixed:
attahtopr

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 1]
```diff
@@ -686,16 +686,27 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
...
+      auto original_methodcall_opcode = pInstr->m_opcode;
+      pInstr->m_opcode = CEE_NOP;
 
+      // replace with a non-virtual call (CALL) to the instrumentation wrapper
       // always use CALL because the wrappers methods are all static
-      pInstr->m_opcode = CEE_CALL;
-      // replace with a call to the instrumentation wrapper
-      pInstr->m_Arg32 = wrapper_method_ref;
+      rewriter_wrapper.CallMemberAfter(wrapper_method_ref, false);
+      rewriter_wrapper.SetILPosition(pInstr->m_pNext);
```
Comment (bobuva) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: @zacharycmontoya Can you clarify the use of `CallMemberAfter` here? To me, it reads like:

1. current instruction is a NOP
2. calling wrapper after the current instruction
3. setting up the arguments last.

Shouldn't the args be set up first?

Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: That's correct, they do need to be set up first and that happens here, maybe in a not obvious way. The `ILInstr` structs are set up as linked lists, so this is what's happening in those lines of code:
1. We set the current instruction to NOP, followed by the call instruction. Result: `nop->call`.
1. We set the "current" instruction to the call instruction. Result `nop->call`
1. Each `LoadInt32` and `LoadInt64` insert BEFORE the current instruction. After the first LoadInt32 which loads the original opcode, the result is: `nop->loadint_opcode->call`
1. We keep inserting load instructions before the call instruction, so the final result is: `nop->loadint_opcode->loadint_mdtoken->loadint_mvid->call`

Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: ok got it. Thanks!

Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: I didn't see that the `ILRewriterWrapper` was putting `Load*` instructions in with `InsertBefore`.

**Pull request merged**